lib/internal/Magento/Framework/Module/Registrar.php

```diff
@@ -32,6 +32,11 @@ public static function registerModule($moduleName, $path)
         self::$modulePaths[$moduleName] = $path;
     }
 
+    public static function unregisterModule($moduleName)
```

Comment:
I'm absolutely with you that Immutability is a good thing.  
However, testability trumps this in my view, and since we are working with global state here, some mutability is probably required.   
It should be possible to create Integration tests without mocks that run in isolation.

I did attempt to resolve the situation without changing the registrar. At first I tried the obvious solution: to run the test in a separate process with the `@runTestsInSeparateProcesses` annotation. But the `\Magento\TestFramework\Isolation\DeploymentConfig` listener sees the registered module as a deployment configuration change and complains:
```
ERROR: deployment configuration is corrupted. The application state is no longer valid.
Further tests may fail. This test failure may be misleading, if you are re-running it on a corrupted application.
```
Next try, use the `@magentoAppIsolation enabled` annotation. With this annotation the listener `\Magento\TestFramework\Annotation\AppIsolation` tries to reset the whole application state after a test with that annotation as follows:
```php
    /**
     * Reset application global state
     */
    protected function _resetApp()
    {
        /** @var $objectManager \Magento\TestFramework\ObjectManager */
        $objectManager = \Magento\TestFramework\Helper\Bootstrap::getObjectManager();
        $objectManager->clearCache();

        \Magento\Framework\Data\Form::setElementRenderer(null);
        \Magento\Framework\Data\Form::setFieldsetRenderer(null);
        \Magento\Framework\Data\Form::setFieldsetElementRenderer(null);
        $this->_appArea = null;
    }
```
This fails since it resets all global state and recreates the application, but it doesn't reset the module registry.

Thinking further, the addition of the `unregisterModule()` method does not conflict with your #1562. In fact, it complements it nicely. 
Once the registrar is instantiated by the object manager as a shared instance, the list of modules in immutable. However, this instance is discarded when the application is reset by `$objectManager->clearCache();`.
Then, when the application is initialized again, a new instance of the registrar is created. This time it will use the changed list of modules.
A test that requires changing the module list state could do the following:
```php
    protected function setUp()
    {
        \Magento\Framework\Module\Registrar::registerModule('Test_Module', BP . '/vendor/test/example');
        \Magento\TestFramework\Helper\Bootstrap::getInstance()->reinitialize();
    }

    protected function tearDown()
    {
        \Magento\Framework\Module\Registrar::unregisterModule($this->vendorModuleName);
        \Magento\TestFramework\Helper\Bootstrap::getInstance()->reinitialize();
        // Alternatively the tests should use the `@magentoAppIsolation` enabled annotation.
    }

```
I've merged your PR #1562 locally and tried this and it works as expected.